### PR TITLE
[mtl] manually retain command buffer data

### LINKED ISSUE #1779: Use non-retaining command buffers on Metal
From https://developer.apple.com/documentation/metal/mtlcommandqueue/1508684-makecommandbufferwithunretainedr :
>Use this method only in extremely performance-critical applications where the cost of maintaining references to objects is significant, and references by some other part of the application can be guaranteed until the object is no longer needed for command execution.

So one important point is Apple did go this extra step to provide a method, meaning that the performance difference is expected. Since we are manually tracking the lifetimes on other backends, we should be using this code paths on Metal, minus some cases where we currently rely on tracking:
  - buffer updates of #1777

### PR DESCRIPTION
**Pull request by kvark**
Fixes #1779
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: Metal

r? @JohnColanduoni @grovesNL 
As mentioned on gitter, this doesn't allow us to get rid of the autorelease pools unfortunately. We can see it as a different issue.

### CODE REVIEW

**Comment by kvark** [comment]
also r? @scoopr

File: src/backend/metal/src/command.rs [hunk 1]
```diff
@@ -751,18 +754,25 @@ impl RawCommandQueue<Backend> for CommandQueue {
...
             }
             command_buffer.commit();
         }
+
+        if let Some(ref fence) = fence {
+            let command_buffer = self.0.new_command_buffer_ref();
+            let value_ptr = fence.0.clone();
+            let fence_block = ConcreteBlock::new(move |_cb: *mut ()| -> () {
+                *value_ptr.lock().unwrap() = true;
+            }).copy();
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: Not part of your change, but why is the `copy` needed here?

File: src/backend/metal/src/command.rs [hunk 2]
```diff
@@ -751,18 +754,25 @@ impl RawCommandQueue<Backend> for CommandQueue {
...
-            // only append the fence handler to the last buffer
-            if i + 1 == num_buffers {
-                if let Some(ref fence) = fence {
-                    let value_ptr = fence.0.clone();
-                    let fence_block = ConcreteBlock::new(move |_cb: *mut ()| -> () {
-                        *value_ptr.lock().unwrap() = true;
-                    }).copy();
-                    msg_send![command_buffer, addCompletedHandler: fence_block.deref() as *const _];
-                }
+            if !inner.retained_buffers.is_empty() {
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: If it's deferred, doesn't this mean we will drop `retained_buffers` on first usage (i.e. application may not keep calling `update_buffer`)?

File: src/backend/metal/src/command.rs (hunk 1 above)
Comment (kvark) [reply src/backend/metal/src/command.rs]: uh, seems totally out of place

File: src/backend/metal/src/command.rs (hunk 2 above)
Comment (kvark) [reply src/backend/metal/src/command.rs]: Great eye! you spotted a bug :) lemme address it...

**Comment by kvark** [comment]
@grovesNL the re-usable command buffers should be addressed now.
The `copy()` on `ConcreateBlock` appears to be needed though, it crushes without it, but I didn't investigate further. Perhaps @JohnColanduoni recalls?..
Anyhow, PTAL.

File: src/backend/metal/src/command.rs [hunk 3]
```diff
@@ -736,13 +742,22 @@ impl RawCommandQueue<Backend> for CommandQueue {
...
+                        let retained_buffers = mem::replace(&mut inner.retained_buffers, Vec::new());
+                        let release_block = ConcreteBlock::new(move |_cb: *mut ()| -> () {
+                            let _ = retained_buffers; // move and auto-release
+                        }).copy();
+                        let cb_ref: &metal::CommandBufferRef = cmd_buffer;
+                        msg_send![cb_ref, addCompletedHandler: release_block.deref() as *const _];
+                    }
+                    cmd_buffer
+                }
+                CommandSink::Deferred { ref passes, .. } => {
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: Doesn't `update_buffer` still cause `retained_buffers` to become populated if the application is currently recording into `CommandSink::Deferred`? Or is this avoided somehow?

Comment (kvark) [reply src/backend/metal/src/command.rs]: Correct, but for multi-use buffer (which is what `Deferred` is for) we don't want to set up the completion handler - retained buffer vectors is going to be cleaned up on `reset()`

Comment (grovesNL) [reply src/backend/metal/src/command.rs]: Right, got it, thanks. Updated commit looks good to me then!

**Comment by kvark** [comment]
Ok, thanks for a good review!
Bors r=grovesNL

**Pull request merged**